### PR TITLE
feat: lock metamask red packet

### PR DESCRIPTION
**Pull request by zhouhanseng**
@guanbinrui

### CODE REVIEW

File: packages/maskbook/src/web3/types.ts [hunk 1]
```diff
@@ -8,6 +10,10 @@ export enum CurrencyType {
     USD = 'usd',
 }
 
+export interface MetaMaskInpageProvider extends ReturnType<typeof createMetaMaskProvider> {
```
Comment (guanbinrui) [inline packages/maskbook/src/web3/types.ts:13]: I think `src/env.d.ts` is a better place for this.

File: packages/maskbook/src/plugins/Wallet/database/types.ts [hunk 2]
```diff
@@ -23,6 +24,7 @@ export interface WalletRecord {
     erc20_token_whitelist: Set<string>
     /** A list of untrusted ERC20 token address */
     erc20_token_blacklist: Set<string>
+    provider: ProviderType
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Wallet/database/types.ts:27]: Did you handle backward compatibility?

**Review by Jack-Works — APPROVED** [review]

File: packages/maskbook/src/plugins/Wallet/database/types.ts (hunk 2 above)
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Wallet/database/types.ts:27]: There should be no `provider` property for WalletRecord anymore. If you wanna get the provider type of the currently selected wallet plz use `currentSelectedWalletProviderSettings` in `packages/maskbook/src/plugins/Wallet/settings.ts`.

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx [hunk 3]
```diff
@@ -130,6 +144,9 @@ export function RedPacket(props: RedPacketProps) {
     const account = useAccount()
     const chainId = useChainId()
     const chainIdValid = useChainIdValid()
+    const wallet = useWallet(payload.sender.address)
+    const isMetamaskRedpacketLocked =
+        useValueRef(currentIsMetamaskLockedSettings) && wallet?.provider === ProviderType.MetaMask
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx:149]:
```suggestion

        useValueRef(currentIsMetamaskLockedSettings) && useValueRef(currentSelectedWalletProviderSettings)=== ProviderType.MetaMask
```

File: packages/maskbook/src/extension/background-script/EthereumServices/providers/MetaMask.ts [hunk 4]
```diff
@@ -6,17 +6,19 @@ import createMetaMaskProvider from 'metamask-extension-provider'
...
+    currentIsMetamaskLockedSettings,
 } from '../../../../plugins/Wallet/settings'
 
-let provider: ReturnType<typeof createMetaMaskProvider> | null = null
+let provider: MetaMaskInpageProvider | null = null
 let web3: Web3 | null = null
 
 async function onAccountsChanged(accounts: string[]) {
     await updateWalletInDB(first(accounts) ?? '')
+    currentIsMetamaskLockedSettings.value = !Boolean(await provider!._metamask?.isUnlocked()) && accounts.length === 0
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/background-script/EthereumServices/providers/MetaMask.ts:21]:
```suggestion
    currentIsMetamaskLockedSettings.value = !(await provider!._metamask?.isUnlocked()) && accounts.length === 0
```

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx [hunk 5]
```diff
@@ -203,6 +220,21 @@ export function RedPacket(props: RedPacketProps) {
...
+            <Card
+                className={classNames(classes.root, {
+                    [classes.metamaskContent]: true,
+                    [classes.cursor]: true,
+                })}
+                onClick={() => Services.Ethereum.connectMetaMask()}
+                component="article"
+                elevation={0}>
+                <MetaMaskIcon className={classes.icon} viewBox="0 0 45 45" />
+                {t('metamask_unlock')}
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx:234]:
```suggestion
                {t('`plugin_wallet_metamask_unlock `')}
```

File: packages/maskbook/src/plugins/Wallet/database/types.ts (hunk 2 above)
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/Wallet/database/types.ts:27]: I Figure it out！

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx (hunk 3 above)
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx:149]: No, you can't do this, this violates the React hooks rule.

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacket.tsx:149]: Notice

File: packages/maskbook/src/web3/types.ts (hunk 1 above)
Comment (zhouhanseng) [reply packages/maskbook/src/web3/types.ts:13]: I think this is web3 related type.

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**